### ISSUE TITLE
twisted.protocols.tls has no producer tests (and producer support may be broken)

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5093]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#5093|
|Type|release blocker: regression|
|Created|2011-05-07 13:58:42Z|

While working on [#7153](https://github.com/twisted/twisted/issues/7153), I noticed some problems with producer/TLS integration. Lack of tests for this in twisted.protocols.tls makes it harder to figure out where the real problem is; the use of a buffering layer within the TLS code suggests that such tests would be useful.

See also [#5063](https://github.com/twisted/twisted/issues/5063).

<details><summary>Searchable metadata</summary>

```
trac-id__5093 5093
type__release_blocker__regression release blocker: regression
reporter__itamar itamar
priority__high high
milestone__Twisted_11_1 Twisted-11.1
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1304776722000000 1304776722000000
changetime__1311596604000000 1311596604000000
version__None None
owner__itamar itamar

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_1"></a>|@itamarst commented|
|-|-|

One problem in particular worth noting: producer pauseProducing notification can take a very long time (e.g. 10 seconds) whereas TCP or old SSL code notifies after far less than a second.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph commented|
|-|-|

If this is true it's definitely a regression.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_3"></a>|@itamarst set owner to @itamarst|
|-|-|

This will be fixed as part of [#5063](https://github.com/twisted/twisted/issues/5063).

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_4"></a>|@itamarst commented|
|-|-|

The pauses mentioned in the comment above were not a real bug. However, producers basically didn't work at all, which is rather more significant. The branch in [#5063](https://github.com/twisted/twisted/issues/5063) makes producers work, and adds test coverage.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_5"></a>|@itamarst set status to `closed`|
|-|-|

(In [32362]) Merge tls-producer-5063-2.

Fixes: [#5063](https://github.com/twisted/twisted/issues/5063), [#5093](https://github.com/twisted/twisted/issues/5093)
Author: itamar, exarkun
Reviewer: exarkun

twisted.protocols.tls (and therefore new reactor TLS codepath) now supports producers correctly.
